### PR TITLE
[MODINVSTOR-485, MODINVSTOR-488] Implement normalized call number index for holdings and item. 

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/MODINVSTOR-485 - BE: Search. Holdings. Search option for Call numbers, normalized (first iteration).
https://issues.folio.org/browse/MODINVSTOR-488 - BE: Search. Item segment. Search option for Effective call numbers (item), normalized (1st iteration)

### Changes:
* New FT `item/holdingsRecord.callNumberAndSuffixNormalized` index;
* New FT `item/holdingsRecord.fullCallNumberNormalized` index.

### CODE REVIEW

File: src/test/java/org/folio/rest/support/HttpClient.java [hunk 1]
```diff
@@ -29,7 +30,7 @@
   private final io.vertx.core.http.HttpClient client;
 
   public HttpClient(Vertx vertx) {
-    client = vertx.createHttpClient();
+    client = vertx.createHttpClient(new HttpClientOptions().setMaxPoolSize(20));
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/support/HttpClient.java]: This is not needed anymore, reverted it.

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+  protected abstract List<String> searchByCallNumberNormalized(String callNumber) throws Exception;
+
+  @Parameters({
+    "germ 350",
+    "germ350",
+    "Germ 350/"
+  })
+  @Test
+  public void canApplyRightTruncationForCallNumber(String searchQuery) throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:27]:
```suggestion
  public void appliesRightTruncationToSearchQuery(String searchQuery) throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    "dictionary"
+  })
+  @Test
+  public void cannotSearchBySuffixOnly(String searchQuery) throws Exception {
+    assertThat(searchByCallNumberNormalized(searchQuery), hasSize(0));
+  }
+
+  protected static String[][] callNumberData() {
+    // Prefix, call number, suffix
+    return new String[][]{
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:130]:
```suggestion
    return new String[][] {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.List;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import junitparams.JUnitParamsRunner;
+import junitparams.Parameters;
+
+@RunWith(JUnitParamsRunner.class)
+public abstract class CallNumberNormalizedTestBase extends TestBaseWithInventoryUtil {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:17]: I like the idea of consolidating much of this into a single suite of tests.

File: src/main/resources/templates/db_scripts/schema.json [hunk 5]
```diff
@@ -640,6 +640,18 @@
           "caseSensitive": false,
           "removeAccents": false
         }
+      ],
+      "fullTextIndex": [
+        {
+          "fieldName": "callNumberAndSuffixNormalized",
+          "sqlExpression" : "normalize_holdings_call_number_and_suffix(jsonb)",
+          "sqlExpressionQuery": "normalize_call_number_string($) || '*'"
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/resources/templates/db_scripts/schema.json:648]: It means right truncation.

File: src/main/resources/templates/db_scripts/createCallNumberNormalizationFunctions.sql [hunk 6]
```diff
@@ -0,0 +1,48 @@
+-- this function is used for call number string and call number search value normalization.
+-- It removes all non-alphanumeric symbols.
+CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.normalize_call_number_string(callNumberString text)
+RETURNS text AS $$
+    SELECT regexp_replace(lower(callNumberString), '[^a-z0-9]', '', 'g');
+$$ LANGUAGE sql IMMUTABLE PARALLEL SAFE STRICT;
+
+-- Holdings call number normalization functions
+
+-- Suffix + call number + prefix normalization. Nulls are omitted.
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/resources/templates/db_scripts/createCallNumberNormalizationFunctions.sql]: Yes, fixed.

**Comment by marcjohnson-kint** [comment]
I appreciate that the tests have been split up into areas of behaviour with clear names, that helps my understanding.

I think, in general, I find the tests hard to follow because the setup (initial state) of the system is separated from the tests themselves, which I find means there is a need to check two places to understand each test.

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    assertThat(callNumbersFound, hasSize(3));
+    assertThat(callNumbersFound, allOf(
+      hasItem("H Germ 350/35: 1"),
+      hasItem("H Germ 350/35: 2"),
+      hasItem("F Germ 350/1")
+    ));
+  }
+
+  @Test
+  public void canApplyRightTruncationForSuffix() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:39]: Does this test demonstrate that the search term can also find matches when it includes part of the suffix?

```suggestion
  public void canSearchByCallNumberAndPartOfSuffix() throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+  protected abstract List<String> searchByCallNumberNormalized(String callNumber) throws Exception;
+
+  @Parameters({
+    "germ 350",
+    "germ350",
+    "Germ 350/"
+  })
+  @Test
+  public void appliesRightTruncationToSearchQuery(String searchQuery) throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:27]:
```suggestion
  public void canSearchByPrefixAndPartOfCallNumber(String searchQuery) throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  @Test
+  public void canApplyRightTruncationForSuffix() throws Exception {
+    final List<String> callNumbersFound = searchByCallNumberNormalized("GE77 .F73 2014 Curriculum");
+
+    assertThat(callNumbersFound, hasSize(1));
+    assertThat(callNumbersFound, hasItem("GE77 .F73 2014 Curriculum Materials Collection"));
+  }
+
+  @Test
+  public void canIgnoreNonAlphanumericCharacters() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:47]:
```suggestion
  public void searchIgnoresNonAlphanumericCharactersInCallNumber() throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    assertThat(callNumbersFound, allOf(
+      hasItem("1—2"),
+      hasItem("1 . 2"),
+      hasItem("1 . . 2"),
+      hasItem("1— —2")
+    ));
+  }
+
+  @Test
+  public void canSkipPrefix() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:60]:
```suggestion
  public void canSearchByCallNumberWithoutPrefix() throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    final List<String> callNumbersFound = searchByCallNumberNormalized("ABC123.1 R15 2018");
+
+    assertThat(callNumbersFound, hasSize(1));
+    assertThat(callNumbersFound, hasItem("Oversize ABC123.1 .R15 2018"));
+  }
+
+  @Test
+  public void canSkipNonAlphanumericCharactersInPrefix() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:68]:
```suggestion
  public void searchIgnoresNonAlphanumericCharactersInPrefix() throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    final List<String> callNumbersFound = searchByCallNumberNormalized("Oversize BX1935.A23 1959");
+
+    assertThat(callNumbersFound, hasSize(1));
+    assertThat(callNumbersFound, hasItem("++ Oversize BX1935 .A23 1959"));
+  }
+
+  @Test
+  public void canSearchByCallNumberAndIgnorePrefixAndSuffix() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:76]:
```suggestion
  public void canSearchByCallNumberWithoutPrefixAndSuffix() throws Exception {
```

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    assertThat(searchByCallNumberNormalized("++"), hasSize(0));
+  }
+
+  @Parameters({
+    "Rare Books S537 N56 C82",
+    "Rare Books S537 N56",
+    "rarebookss537n56c82",
+  })
+  @Test
+  public void canSearchByPrefixAndCallNumber(String searchQuery) throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:97]: How is this different to the first test?

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java (hunk 7 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:39]: Yes, the test demonstrates that if user entered call number + part of the suffix, then right truncation for the suffix should be applied.

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java (hunk 13 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:97]: In this case call numbers also have a suffix.

Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java:97]: Can the tests be renamed to reflect that difference? For example, CanSearchByPrefixAndCallNumberEvenWhenMatchesHaveSuffixes (I'm not super keen on this name, hopefully you can this of a better one)

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java (hunk 7 above)
Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java]: Ok, thanks, I wanted to confirm my assumption to go with my suggestion

File: src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java (hunk 13 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/CallNumberNormalizedTestBase.java]: Renamed.
